### PR TITLE
Touch Compatibility to CircuitVerse -ZadeAbhishek || GSoC'21

### PR DESCRIPTION
**Pull request by ZadeAbhishek**
### **Feature**
The PR is for **touch compatibility to CircuitVerse**
CircuitVerese just supports the Mouse on the browser till now. In this PR the website will support both Touch and mouse on Brower simultanously. 

This is Done by adding touch listeners

- 1. touchstart
- 2. touchmove
- 3. touchend

#### Describe the changes you have made in this PR -

1. Added **touch listeners** in listeners.js
2. Function for pan on the simulator is made common for both touch and mouse in listerners.js

3.**Pinch to Zoom** is implemented to zoom in and zoom out on the simulator using pinch.

Note: Please check **Allow edits from maintainers.** if you would like us to assist in the PR.

### CODE REVIEW

**Review by ShreyaPrasad1209 — CHANGES_REQUESTED** [review]
@ZadeAbhishek great work! Please look into the changes suggested by Code Climate. In the meanwhile, I'll review your code.

**Comment by ZadeAbhishek** [comment]
> @ZadeAbhishek great work! Please look into the changes suggested by Code Climate. In the meanwhile, I'll review your code. 

On it!!

**Comment by abstrekt** [comment]
Hey, @ZadeAbhishek first of all nice work!
I have tested out these changes and here are a few things I noticed:
* Firstly - zooming on the canvas and zooming the UI are two different things - do keep note of this when you start working on the UI cause this thing is very "interfering" when using a touch only device.
1. When I select an element from the panel for placing, it is immediately seen on top of the LAST PLACED element, 
image
https://user-images.githubusercontent.com/42478217/122637910-c896b780-d10e-11eb-96b0-40fc25d4db5b.mov
though we can place the element wherever we want by clicking on the desired position on the canvas after selection, can this immediate placement be avoided? also at some times I'm no longer able to place elements < unable to reproduce  - just noting it.
2. Connection of wires seems to be working fine 🎉
3. Unable to select an element :( , we need to be able to select an element on touch,  for things like dragging, changing bits, etc.
4. I see that we can touch and pan the canvas, but also saw I was able to do a areal selection (like on windows desktop) but don't know how it was activated?
#### DO ONE THING: Make a doc, write all the test cases, and usabilities (though the UX should be made such that it itself tells the user on how-to-use-me)
5. Current zoom levels of the canvas works fine for desktop screens, but for a mobile screen that zoom level is too low to work on. We should make elements and canvas feel native to the device (perspective), it should not feel similar to how we feel when we (View as Destop Site on mobile browsers). 
6. Two figner zoom originates from the center of canvas, we should make it able to zoom as the center of the two fingers as the origin.

* If you haven't I recommend to test all elements's usabilities, look out for any barriers while using any element on touch, such as the imageAnnotation element, are we handling double click on touch?
image

File: simulator/src/embedListeners.js [hunk 1]
```diff
@@ -1,104 +1,216 @@
...
+    updatePositionSet(true);
+    updateCanvasSet(true);
+
+    simulationArea.lastSelected = undefined;
+    simulationArea.selected = false;
+    simulationArea.hover = undefined;
+    const rect = simulationArea.canvas.getBoundingClientRect();
+    simulationArea.mouseDownRawX = (embedCoordinate.x - rect.left) * DPR;
+    simulationArea.mouseDownRawY = (embedCoordinate.y - rect.top) * DPR;
+    simulationArea.mouseDownX = Math.round(((simulationArea.mouseDownRawX - globalScope.ox) / globalScope.scale) / unit) * unit;
```
Comment (ShreyaPrasad1209) [inline simulator/src/embedListeners.js:39]: Write the meaning of ox and oy in comments

**Review by ShreyaPrasad1209 — CHANGES_REQUESTED** [review]
Great work @ZadeAbhishek!🎉 I have suggested a few changes, after which we can wait for @abstrekt, @satu0king and @nitin10s’s review.

**Review by ShreyaPrasad1209 — APPROVED** [review]

**Review by abstrekt — APPROVED** [review]
> Hey, @ZadeAbhishek first of all nice work!
> I have tested out these changes and here are a few things I noticed:
> 
> * Firstly - zooming on the canvas and zooming the UI are two different things - do keep note of this when you start working on the UI cause this thing is very "interfering" when using a touch only device.
> 
> 1. When I select an element from the panel for placing, it is immediately seen on top of the LAST PLACED element,
>    image
>    https://user-images.githubusercontent.com/42478217/122637910-c896b780-d10e-11eb-96b0-40fc25d4db5b.mov
>    though we can place the element wherever we want by clicking on the desired position on the canvas after selection, can this immediate placement be avoided? also at some times I'm no longer able to place elements < unable to reproduce  - just noting it.
> 2. Connection of wires seems to be working fine 🎉
> 3. Unable to select an element :( , we need to be able to select an element on touch,  for things like dragging, changing bits, etc.
> 4. I see that we can touch and pan the canvas, but also saw I was able to do a areal selection (like on windows desktop) but don't know how it was activated?
> 
> #### DO ONE THING: Make a doc, write all the test cases, and usabilities (though the UX should be made such that it itself tells the user on how-to-use-me)
> 1. Current zoom levels of the canvas works fine for desktop screens, but for a mobile screen that zoom level is too low to work on. We should make elements and canvas feel native to the device (perspective), it should not feel similar to how we feel when we (View as Destop Site on mobile browsers).
> 2. Two figner zoom originates from the center of canvas, we should make it able to zoom as the center of the two fingers as the origin.
> 
> * If you haven't I recommend to test all elements's usabilities, look out for any barriers while using any element on touch, such as the imageAnnotation element, are we handling double click on touch?
>   image


There still seems to be a bit of inconsistency on point 1, though elements are not getting place on top of the last placed element , they are getting placed randomly to the right of the canvas, something too far away, or sometimes less, I guess the plan was to place it on the right of the last placed element? --  distance is varying,


rest seems to be workable, though performance seems to be a concern here on my device, if doing basic things like dragging the element, panning, placing elements etc causing lag, seems it'll be hard to use it for the average users 🤔

**Comment by abstrekt** [comment]
we can always come back and improve 👍🏻

**Comment by ZadeAbhishek** [comment]
@abstrekt have you tested on landscape or Portrait mode ? My observation is In Landscape is works fine but in portrait mode it lags.

**Comment by abstrekt** [comment]
@tachyons  I guess we can merge this, there are a few bugs here and there but those can be fixed later on.

**Comment by tachyons** [comment]
@ZadeAbhishek @abstrekt This is no longer needed, right?

**Comment by ZadeAbhishek** [comment]
> @ZadeAbhishek @abstrekt This is no longer needed, right?

not needed

**Pull request closed without merging**